activitysim/abm/models/tour_mode_choice_reallocation.py

```diff
@@ -0,0 +1,653 @@
+
+    Parameters
+    ----------
+    tours : pandas.DataFrame
+    segment_column_name : str
+        column in tours table used for segmenting model spec
+    model_settings : dict
+    trace_label : str
+
+    Returns
+    -------
+    tours : pd.DataFrame
+        Adds two * n_modes logsum columns to each tour row, e.g. "logsum_DRIVE_outbound"
+    """
+
+    # create pseudo-trips from tours for all tour modes
+    logsum_trips = create_logsum_trips(
+        tours, segment_column_name, model_settings, trace_label
+    )
+
+    # temporarily register trips in the pipeline
+    pipeline.replace_table("trips", logsum_trips)
+    tracing.register_traceable_table("trips", logsum_trips)
+    pipeline.get_rn_generator().add_channel("trips", logsum_trips)
+
+    # run trip mode choice on pseudo-trips. use orca instead of pipeline to
+    # execute the step because pipeline can only handle one open step at a time
+    orca.run(["trip_mode_choice"])
+
+    # add trip mode choice logsums as new cols in tours
+    tours = append_tour_leg_trip_mode_choice_logsums(tours)
+
+    # de-register logsum trips table
+    pipeline.get_rn_generator().drop_channel("trips")
+    tracing.deregister_traceable_table("trips")
+
+    return tours
+
+
+def code_household_auto_sufficiency(households):
+    """
+    Adds and codes household auto sufficiency based on household auto ownership and number of drivers in the household
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+
+    Returns
+    -------
+    households : pd.Dataframe
+        Household table with auto sufficiency column added
+
+    """
+    households['auto_sufficiency'] = 'suff'
+    households.loc[
+        households.auto_ownership.between(0, households.num_drivers, 'neither'), 'auto_sufficiency'] = 'insuff'
+    households.loc[households.auto_ownership == 0, 'auto_sufficiency'] = 'nocar'
+
+    return households
+
+
+def filter_for_multi_drive_tour_households(households, persons, tours):
+    """
+    Filters tours down to only those from auto insufficient households with multiple drivers who made SOV-based tours
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+    persons : pd.Dataframe
+    tours : pd.Dataframe
+
+    Returns
+    -------
+    tours_insuff_mult_drv : pd.Dataframe
+        Tours from auto insufficient households with multiple drivers who made SOV-based tours
+
+    """
+    households = code_household_auto_sufficiency(households)
+
+    # Isolate for SOV-based tours by driving aged persons in auto insufficient households.
+    households_insuff = households[households.auto_sufficiency == 'insuff']
+    person_drivers = persons[persons.is_driving_age]
+    # TODO: Expand criteria to cover joint tours once a reallocation method can be developed for it
+    tours_insuff_drv = tours[
+        tours.household_id.isin(households_insuff.index) &
+        tours.person_id.isin(person_drivers.index) &
+        (tours.tour_mode == 'DRIVEALONEFREE')]
```

Comment:
@TianYangLin We should probably also look for `DRIVEALONEPAY` for the sake of completeness, even if it's not used in PTM

Reply:
Will add DRIVEALONEPAY